Invalidate smart group cache for group following deletion of group_c…

seamuslee001 opened the pull request:
…ontact row

Overview
----------------------------------------
This is a follow on from @eileenmcnaughton 's work to ensure that when contacts are removed from a smart group they are removed from the cache instantly, This follows a similar principle however in this case what we are doing is removing a row such as saying a contact that would otherwise be in a smart group is removed. If that row is deleted from the civicrm_group_contact table we should ensure the cache gets updated

Before
----------------------------------------
Contact maybe back in smart group but not until cache is refreshed. 

After
----------------------------------------
Cache updated immediately

ping @mattwire

eileenmcnaughton commented:
I see the logic but there is a performance risk here - still I think we are only doing this when we 'remove a removal' - which seems like a rare action - I'm not even 100% sure I know how to do that via the UI.

I will wait for @mattwire to weigh in...

seamuslee001 commented:
@eileenmcnaughton to do this in the UI go to the groups tab in a contact record look under removed groups you will have rejoin (update status in civicrm_group_contact to be Added) or delete

eileenmcnaughton commented:
That feels suitably obscure that the performance hit feels ok!

eileenmcnaughton commented on CRM/Contact/BAO/GroupContact.php:
```diff
@@ -211,6 +213,12 @@ public static function removeContactsFromGroup(
         ];
         CRM_Contact_BAO_SubscriptionHistory::create($historyParams);
+        // Removing a row from civicrm_group_contact for a smart group may mean a contact
+        // Is now back in a group based on criteria so we should re-load the cache.
```
Actually it's never a good idea to pro-actively load a cache in my experience - it will get loaded if needed

mattwire commented on CRM/Contact/BAO/GroupContact.php:
```diff
@@ -211,6 +213,12 @@ public static function removeContactsFromGroup(
+        $groupInformation = civicrm_api3('Group', 'getsingle', ['id' => $groupId]);
+        if (!empty($groupInformation['saved_search_id'])) {
+          CRM_Contact_BAO_GroupContactCache::load($groupId, TRUE);
+        }
```
@seamuslee001 Instead of explicitly loading the cache here how about adding a new function to `CRM_Contact_BAO_GroupContactCache` that invalidates the cache for that group (meaning that it will be rebuilt next time it is needed):
```
public static function expireGroupContactCache($groupID) {
  $update = "
  UPDATE civicrm_group g
  SET    cache_date = null, refresh_date = null
  WHERE  id = %1 ";

  $params = [
    1 => [$groupID, 'Integer'],
  ];

  CRM_Core_DAO::executeQuery($query, $params);
}
```

That should cause the cache for that group to be rebuilt next time it is accessed.  And I think we could start using the function in a lot more places to stop the cache being rebuilt just because we've made a change, rather than because we actually need to use it!

eileenmcnaughton replied:
yes I agree if we invalidate rather than flush it the cache then it could be cleared by the relevant scheduled job with less uI impact

seamuslee001 commented:
@eileenmcnaughton @mattwire i have now added the function as per Matt to invalidate the group contact cache and updated appropriately, is this good now?

eileenmcnaughton commented:
Hah - now it's your style jenkins dislikes. Doesn't feel so fun where you're on the receiving end does it now..... :-)

eileenmcnaughton commented:
I agree with the fix as is and for what it is worth I do like your style!

Get jenkins to agree & I'm in.

@davejenx @jaapjansma @bjendres pinging you as I think you have an interest in this area & I think it's a good technique

eileenmcnaughton commented on CRM/Contact/BAO/GroupContact.php:
```diff
@@ -211,6 +213,13 @@ public static function removeContactsFromGroup(
+        // So that accurate group cache is created next time it is needed.
+        $groupInformation = civicrm_api3('Group', 'getsingle', ['id' => $groupId]);
+        if (!empty($groupInformation['saved_search_id'])) {
+          CRM_Contact_BAO_GroupContactCache::invalidateGroupContactCache($groupId);
```
Actually don't check for saved search id here - it is also relevant if a group has children and I don't think it would be BAD to do if not relevant

mattwire replied:
@eileenmcnaughton agree

mattwire commented:
@seamuslee001 Thankyou - looks good.  Fix the style issues and remove the saved_search check per @eileenmcnaughton comments and this is good to merge.

seamuslee001 commented:
done @mattwire @eileenmcnaughton

eileenmcnaughton commented:
great - thanks @seamuslee001
